Annotated functional interfaces with @FunctionalInterface.

ccreeten opened the pull request.

jvdb commented:
`ParseItem` is not a functional interface. You can implement it by overriding one method, but it would give you a useless `ParseItem` if you don't also override one `isX()` and one `asX()` method.

Furthermore, it's important to only annotate interfaces that are actually desired to be a functional interface. For example, `ByteStream` is rarely implemented as a lambda in a realistic setting.

ccreeten commented:
> ParseItem is not a functional interface. You can implement it by overriding one method, but it would give you a useless ParseItem if you don't also override one isX() and one asX() method.
> 
> Furthermore, it's important to only annotate interfaces that are actually desired to be a functional interface. For example, ByteStream is rarely implemented as a lambda in a realistic setting.

Valid points!

The reason I added them was because my version of the Eclipse compiler had a bug, where it did not detect an interface as functional when dealing with multiple inheritance, where the supers have an abstract and default method with the same signature. This is the case with Trampoline. To be totally sure, I added the annotation to allow for compile time checking and making it very clear they are (and as a result, can be used for lambda's).

Should I decline the pr, or update to only annotate the logical interfaces (trampoline, callbacks, expressions?)?

jvdb commented:
I think it's a good idea to have the annotations, so if you update it by removing the annotations from `ParseItem` and `ByteStream`, I will merge it!

rdvdijk commented:
👍
